crypto_box.js

```diff
@@ -95,6 +103,159 @@ function crypto_box_seal_open (m, c, pk, sk) {
+
+  const k = new Uint8Array(crypto_box_BEFORENMBYTES)
+
+  if (crypto_box_beforenm(k, pk, sk) !== 0) {
+    return -1
+  }
+
+  return crypto_box_detached_afternm(c, mac, m, n, k)
+}
+
+// int
+// crypto_box_easy(unsigned char *c, const unsigned char *m,
+//                 unsigned long long mlen, const unsigned char *n,
+//                 const unsigned char *pk, const unsigned char *sk)
+// {
+//     if (mlen > crypto_box_MESSAGEBYTES_MAX) {
+//         sodium_misuse();
+//     }
+//     return crypto_box_detached(c + crypto_box_MACBYTES, c, m, mlen, n,
+//                                pk, sk);
+// }
+function crypto_box_easy (c, m, n, pk, sk) {
+  assert(c.length <= crypto_box_MESSAGEBYTES_MAX, "m should not be more than 'crypto_box_MESSAGEBYTESMAX' bytes")
+
+  return crypto_box_detached(c + crypto_box_MACBYTES, c, m, n, pk, sk)
+}
+// int
+// crypto_box_open_detached_afternm(unsigned char *m, const unsigned char *c,
+//                                  const unsigned char *mac,
+//                                  unsigned long long clen,
+//                                  const unsigned char *n,
+//                                  const unsigned char *k)
+// {
+//     return crypto_secretbox_open_detached(m, c, mac, clen, n, k);
+// }
+function crypto_box_open_detached_afternm (m, c, mac, n, k) {
+  return crypto_secretbox_open_detached(m, c, mac, n, k)
+}
+
+// int
+// crypto_box_open_detached(unsigned char *m, const unsigned char *c,
+//                          const unsigned char *mac,
+//                          unsigned long long clen, const unsigned char *n,
+//                          const unsigned char *pk, const unsigned char *sk)
+// {
+//     unsigned char k[crypto_box_BEFORENMBYTES];
+//     int           ret;
+//
+//     if (crypto_box_beforenm(k, pk, sk) != 0) {
+//         return -1;
+//     }
+//     ret = crypto_box_open_detached_afternm(m, c, mac, clen, n, k);
+//     sodium_memzero(k, sizeof k);
+//
+//     return ret;
+// }
+function crypto_box_open_detached (m, c, mac, n, pk, sk) {
+  check(mac, crypto_box_MACBYTES)
+  check(n, crypto_box_NONCEBYTES)
+  check(pk, crypto_box_PUBLICKEYBYTES)
+  check(sk, crypto_box_SECRETKEYBYTES)
+
+  const k = Uint8Array(crypto_box_BEFORENMBYTES)
+
+  if (crypto_box_beforenm(k, pk, sk) !== 0) {
+    return -1
+  }
+
+  return crypto_box_open_detached_afternm(m, c, mac, n, k)
+}
+
+// int
+// crypto_box_open_easy(unsigned char *m, const unsigned char *c,
+//                      unsigned long long clen, const unsigned char *n,
+//                      const unsigned char *pk, const unsigned char *sk)
+// {
+//     if (clen < crypto_box_MACBYTES) {
+//         return -1;
+//     }
+//     return crypto_box_open_detached(m, c + crypto_box_MACBYTES, c,
+//                                     clen - crypto_box_MACBYTES,
+//                                     n, pk, sk);
+// }
+function crypto_box_open_easy (m, c, n, pk, sk) {
+  check(m, crypto_box_MACBYTES)
```

Comment:
On second though, this probably isn't a MAC.

Reply:
Yeh I think this should be `check(m, c.byteLength - crypto_box_MACBYTES)`. Similarly the ciphertext length should be checked in `crypto_box_open` and the detached methods if we end up exporting them.